### ISSUE TITLE
Memory leak when loading new drumkits

### ISSUE THREAD

**Issue by mauser** [hydrogen-music/hydrogen#655]
Hydrogen version *          : 1.0-beta1

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

When loading a new drumkit in Hydrogen, it seems that the application does not delete all objects after exiting. Especially all Layers, Components and Samples are not getting deleted properly. 

```
Objects map :                        class	constr   destr   alive
	[            AudioEngineInfoForm ]	     1	     1	     0
	[                       Director ]	     1	     1	     0
	[              SimpleHTMLBrowser ]	     1	     1	     0
	[                    HydrogenApp ]	     1	     1	     0
	[               InstrumentEditor ]	     1	     1	     0
	[          InstrumentEditorPanel ]	     1	     1	     0
	[                   LayerPreview ]	     1	     1	     0
	[                    WaveDisplay ]	     2	     2	     0
	[                 InstrumentRack ]	     1	     1	     0
	[             LadspaFXProperties ]	     4	     4	     0
	[                       MainForm ]	     1	     1	     0
	[                          Mixer ]	     1	     1	     0
	[              DrumPatternEditor ]	     1	     1	     0
	[            NotePropertiesRuler ]	     5	     5	     0
	[    PatternEditorInstrumentList ]	     1	     1	     0
	[             PatternEditorPanel ]	     1	     1	     0
	[             PatternEditorRuler ]	     1	     1	     0
	[                PianoRollEditor ]	     1	     1	     0
	[                  PlayerControl ]	     1	     1	     0
	[                MetronomeWidget ]	     1	     1	     0
	[                 PlaylistDialog ]	     1	     1	     0
	[                     SongEditor ]	     1	     1	     0
	[          SongEditorPatternList ]	     1	     1	     0
	[        SongEditorPositionRuler ]	     1	     1	     0
	[                SongEditorPanel ]	     1	     1	     0
	[           SoundLibraryDatabase ]	     1	     1	     0
	[               SoundLibraryInfo ]	     2	     2	     0
	[              SoundLibraryPanel ]	     1	     1	     0
	[               SoundLibraryTree ]	     1	     1	     0
	[                   SplashScreen ]	     1	     1	     0
	[             AutomationPathView ]	     1	     1	     0
	[                         Button ]	   195	   195	     0
	[                  CpuLoadWidget ]	     1	     1	     0
	[                          Fader ]	    20	    20	     0
	[                    MasterFader ]	     1	     1	     0
	[                           Knob ]	    72	    72	     0
	[                       LCDDigit ]	   452	   452	     0
	[                     LCDDisplay ]	    86	    86	     0
	[                     LCDSpinBox ]	     1	     1	     0
	[                       LCDCombo ]	     5	     5	     0
	[             MidiActivityWidget ]	     1	     1	     0
	[                   PixmapWidget ]	    78	    78	     0
	[                         Rotary ]	    37	    37	     0
	[                AlsaAudioDriver ]	     1	     1	     0
	[                 AlsaMidiDriver ]	     1	     1	     0
	[                  TransportInfo ]	     1	     1	     0
	[                    AudioEngine ]	     1	     1	     0
	[       AutomationPathSerializer ]	     1	     1	     0
	[                           ADSR ]	  2850	  2850	     0
	[                 AutomationPath ]	     1	     1	     0
	[                        Drumkit ]	    46	    46	     0
	[               DrumkitComponent ]	    48	    47	     1
	[                     Instrument ]	  1430	  1430	     0
	[            InstrumentComponent ]	  1441	  1430	    11
	[                InstrumentLayer ]	  4219	  4168	    51
	[                 InstrumentList ]	    47	    47	     0
	[                        Pattern ]	    10	    10	     0
	[                    PatternList ]	     4	     4	     0
	[                       Playlist ]	     1	     1	     0
	[                         Sample ]	  4218	  4167	    51
	[                           Song ]	     1	     1	     0
	[                     SongReader ]	     1	     1	     0
	[           CoreActionController ]	     1	     1	     0
	[                     EventQueue ]	     1	     1	     0
	[                        Effects ]	     1	     1	     0
	[                   LadspaFXInfo ]	    20	    20	     0
	[                        XMLNode ]	 10078	 10078	     0
	[                         XMLDoc ]	    50	    50	     0
	[                       Hydrogen ]	     1	     1	     0
	[                   LocalFileMng ]	     1	     1	     0
	[                     MidiAction ]	  1469	  1469	     0
	[                  ActionManager ]	     1	     1	     0
	[                        MidiMap ]	     1	     1	     0
	[                      NsmClient ]	     1	     1	     0
	[                      OscServer ]	     1	     1	     0
	[                    Preferences ]	     1	     1	     0
	[               WindowProperties ]	    64	    64	     0
	[                        UIStyle ]	     1	     1	     0
	[                     H2RGBColor ]	   117	   117	     0
	[                        Sampler ]	     1	     1	     0
	[                          Synth ]	     1	     1	     0
	[                       Timeline ]	     1	     1	     0
Total :    114 objects.

```

**Comment by mauser** [comment]
It seems that those issues have a very similar source as 7e49f10baed47cc927a4de78d4521714b6224bd1 and related commits. In different locations std::vector.clear() is used, which does not delete it contents. 

Especially for the instrument components, using a std::shared_ptr / unique_ptr might be reasonable, since the overhead should come into play during constructing and copying those pointers, but not while dereferencing.

**Comment by mauser** [comment]
Fixed by b81858685ec7a16e4cfc224e19bbee2772a778eb. Though on the long run, the use of smartptrs might be better.

**Comment by elpescado** [comment]
AFAIK, `std::unique_ptr` does not incur any runtime overhead - all templated code should be inlined anyway.
